File: 27. Producer consumer using queue.py
"""
Task: Producer-consumer simulation using queue and threading.Condition
Time Complexity: Depends on workload
Space Complexity: O(n)
"""

import threading
from collections import deque
import time

q = deque()
cond = threading.Condition()
done = False


def producer():
    global done
    for i in range(1,6):
        with cond:
            q.append(i)
            print('produced', i)
            cond.notify()
        time.sleep(0.1)
    with cond:
        done = True
        cond.notify_all()


def consumer(cid):
    while True:
        with cond:
            while not q and not done:
                cond.wait()
            if q:
                v = q.popleft(); print('consumer',cid,'consumed',v)
            elif done:
                break

if __name__=="__main__":
    t1 = threading.Thread(target=producer); t2 = threading.Thread(target=consumer, args=(1,)); t1.start(); t2.start(); t1.join(); t2.join()

# Task: Producer consumer using queue

File: 27. Producer consumer using queue.cpp
/*
 Task: Producer-consumer simulation using queue and condition_variable
*/

#include <bits/stdc++.h>
using namespace std;

queue<int> q;
mutex m;
condition_variable cv;
bool done = false;

void producer()
{
    for (int i = 1; i <= 5; ++i)
    {
        {
            unique_lock<mutex> lk(m);
            q.push(i);
            cout << "produced " << i << "\n";
        }
        cv.notify_one();
        this_thread::sleep_for(chrono::milliseconds(100));
    }

    {
        unique_lock<mutex> lk(m);
        done = true;
    }
    cv.notify_all();
}

void consumer(int id)
{
    while (true)
    {
        unique_lock<mutex> lk(m);
        cv.wait(lk, [] { return !q.empty() || done; });

        if (!q.empty())
        {
            int v = q.front();
            q.pop();
            cout << "consumer " << id << " consumed " << v << "\n";
        }
        else if (done)
        {
            break;
        }
    }
}

int main()
{
    thread p(producer);
    thread c1(consumer, 1);
    thread c2(consumer, 2);

    p.join();
    c1.join();
    c2.join();

    return 0;
}

// Time Complexity: Depends on workload
// Space Complexity: O(n)
